src/lib/http-observer.js

```diff
@@ -0,0 +1,220 @@
+import Uri from 'urijs';
+
+import Logger from './log';
+import { CDN_HEADERS, CDN_HOSTNAMES } from './cdn';
+
+const logger = new Logger('webext.lib.http-observer', console);
+
+export const REQUEST_TYPES = [
+  'main_frame',
+  'sub_frame',
+  'stylesheet',
+  'script',
+  'image',
+  'object',
+  'xmlhttprequest',
+  'xbl',
+  'xslt',
+  'ping',
+  'beacon',
+  'xml_dtd',
+  'font',
+  'media',
+  'websocket',
+  'csp_report',
+  'imageset',
+  'web_manifest',
+  'other'
+];
+
+export default class HttpObserver {
+  constructor() {
+    this.requests = {};
+    this.addListeners();
+  }
+
+  addListeners() {
+    browser.webNavigation.onBeforeNavigate.addListener(details => {
+      this.navigationListener(details);
+    });
+    browser.webRequest.onBeforeRequest.addListener(
+      details => this.requestListener(details),
+      { urls: [ '<all_urls>' ] }
+    );
+    browser.webRequest.onCompleted.addListener(
+      details => this.responseListener(details),
+      { urls: [ '<all_urls>' ] },
+      [ 'responseHeaders' ]
+    );
+  }
+
+  // Whenever the user navigates to a new page, clear the collected requests
+  // for this tab.
+  navigationListener(details) {
+    this.requests[details.tabId] = {};
+  }
+
+  // Whenever a request is made, log some initial metadata. Only do this if the
+  // instance was listening at webNavigation.onBeforeNavigate.
+  requestListener(details) {
+    if (details.tabId in this.requests) {
+      this.requests[details.tabId][details.requestId] = {
+        start: details.timeStamp,
+        url: details.url,
+        completed: false
+      };
+    }
+  }
+
+  // Whenever a request is completed, log additional metadata. Only do this if
+  // the instance was listening at webNavigation.onBeforeNavigate.
+  responseListener(details) {
+    if (details.tabId in this.requests) {
+      const request = this.requests[details.tabId][details.requestId];
```

Comment:
Also tossed in a log statement for `navigationListener`, since the tab's requests get cleared there. Seems like the process of just loading this page without clicking on anything triggers multiple navigation events:

<img width="1103" alt="screenshot 2017-02-15 17 18 50" src="https://cloud.githubusercontent.com/assets/21687/22997983/d9f4fc24-f3a2-11e6-9128-a07656211e2d.png">